src/container/srv_container.c

```diff
@@ -1427,6 +1427,137 @@ ds_cont_close_by_pool_hdls(uuid_t pool_uuid, uuid_t *pool_hdls, int n_pool_hdls,
 	return rc;
 }
 
+/* argument type for callback function to list containers */
+struct list_cont_iter_args {
+	uuid_t				pool_uuid;
+
+	/* Remaining containers in client buf, total containers in pool */
+	uint64_t			avail_ncont;
+	uint64_t			ncont;
+
+	/* conts[]: capacity, and current index */
+	uint64_t			conts_len;
+	uint64_t			conts_index;
+	struct daos_pool_cont_info	*conts;
+};
+
+/* callback function for list containers iteration. */
+static int
+enum_cont_cb(daos_handle_t ih, d_iov_t *key, d_iov_t *val, void *varg)
+{
+	struct list_cont_iter_args	*ap = varg;
+	struct daos_pool_cont_info	*cinfo;
+	uuid_t				 cont_uuid;
+	(void) val;
+
+	if (key->iov_len != sizeof(uuid_t)) {
+		D_ERROR("invalid key size: key="DF_U64"\n", key->iov_len);
+		return -DER_IO;
+	}
+
+	uuid_copy(cont_uuid, key->iov_buf);
+	ap->ncont++;
+
+	D_DEBUG(DF_DSMS, "pool/cont: "DF_CONTF"\n",
+		DP_CONT(ap->pool_uuid, cont_uuid));
+
+	/* Client didn't provide buffer or not enough space */
+	if (ap->avail_ncont < ap->ncont)
+		return 0;
+
+	/* Realloc conts[] if needed (double each time starting with 1) */
+	if (ap->conts_index == ap->conts_len) {
+		void	*ptr;
+		size_t	realloc_bytes;
+		size_t	realloc_elems = (ap->conts_len == 0) ? 1 :
+					ap->conts_len * 2;
+
+		realloc_bytes = (realloc_elems *
+				sizeof(struct daos_pool_cont_info));
+		D_REALLOC(ptr, ap->conts, realloc_bytes);
```

Comment:
There's also D_REALLOC_ARRAY. But be aware that the D_REALLOC family share one behavior: when growing an allocation, they do not zero the new bytes. I consider this a bug (filed a CART ticket), as this is inconsistent with their D_ALLOC relatives. So make sure you're not relying on them to zero your new bytes.

Reply:
Thanks for that suggestion. I will remove the realloc_bytes variable and use that array macro. I think the code does not depend on zeroing the extended portion of the allocation.